single-registry-architecture/README.md

```diff
@@ -0,0 +1,217 @@
+customer's images in the default configuration. This means, that they can start hitting rate limiting where previously
+nothing like that happened. It also means that they might want to configure their own registries in the container
+runtimes to work around this issue.
+
+Using this approach, we will be able to simplify our architecture, limit the cost (Docker Hub is expensive and gets worse)
+and keep HA properties of our infrastructure.
+
+### Specific problems and goals to solve
+
+More specifically, our aim is to:
+
+- Stop using Docker Hub by the end of 2023.
+- Simplify registries architecture.
+- Switch entirely to OCI based registries for both container images and helm charts.
+  - Flux developers have already declared that no new feature will come to Flux for users of HTTP based repos, only for
+  - OCI. Currently, we use a mix of HTTP and OCI repos, but charts discovery uses HTTP based ones only.
+- If possible, switch to single OCI provider.
+- Increase availability and lower the network transfer cost by providing a pull-through cache service.
+- Automatically clean old dev release artifacts.
+- If possible, use the registry provider to replicate registries between Europe and China.
+- Use a solution, that works natively with `containerd` and `helm` and doesn't need tricks like mutating webhooks,
+  that change Pod's image.
+
+## Proposed solution
+
+The target architecture would look like below (dashed lines for pull actions mean fall-back registries):
+
+![Registries architecture diagram](./2023-10-registries.drawio.svg "Registries architecture")
+
+Please note, that instead of a per-MC cache instance, we might want to run just one cache per cloud provider's region.
+Check the discussion [below](#open-questions).
+
+### Choosing OCI repository provider
+
+We checked for the single best registry and chose ACR. We did some evaluation before and you can read results here:
+
+- <https://github.com/giantswarm/roadmap/issues/2382>
+- <https://hackmd.io/gqF6PT53RcG_i4OhrmtWIw#Note-on-compatibility>
+
+Additionally, ACR offers th following:
+
+- High availability by definition.
+- Automatic geo-replication, including geo-replication between Europe and China (confirmed by Azure support, requires
+  going through a formal process, though).
+- Automatic retention
+  - When enabled, manifests that don't have any associated tags (untagged manifests) and are not locked, will be
+    automatically deleted after the number of retention days specified. [Docs](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-retention-policy).
+- Best-in-class support for the OCI standard (it's MS that is pushing it forward).
+- Reasonable pricing
+  - price per region €1.540/day (500 GB included), then €0.003/GB/day + traffic fee
+  - [Azure cost calculator](https://azure.microsoft.com/en-us/pricing/calculator/?service=container-registry)
+
+### Choosing caching solution
```

Comment:
Hopefully it's not too late :)
have you considered Zot registry? It seems it fits our needs (https://zotregistry.io/v1.4.3/articles/mirroring/)

Reply:
I don't disagree, but ideally there is proper ownership of the component then.  
Whichever team owns it, need to make sure it's running day and night.  

Because of the above, my vote goes for a managed solution.

Reply:
I mean zot for the caching/mirror solution.

Reply:
No, I missed this one somehow. I'll catch up and try to check if and how we can use it for caching.